src/Disks/IO/CachedOnDiskReadBufferFromFile.cpp

```diff
@@ -274,16 +271,7 @@ CachedOnDiskReadBufferFromFile::getReadBufferForFileSegment(FileSegmentPtr & fil
                     return getCacheReadBuffer(range.left);
                 }
 
-                if (wait_download_tries++ < wait_download_max_tries)
-                {
-                    download_state = file_segment->wait();
-                }
-                else
-                {
-                    LOG_DEBUG(log, "Retries to wait for file segment download exceeded ({})", wait_download_tries);
-                    download_state = FileSegment::State::SKIP_CACHE;
-                }
-
+                download_state = file_segment->wait();
```

Comment:
The main point is here. In complete() method we always called notify, but complete() could be called when the state did not change, so we returned from wait() for no reason.